In [1]:
import numpy as np
import matplotlib as plt
from scipy.linalg import lu

from sympy import *
init_printing(latex=True)

<h3>문제 1</h3>

<h5>1번</h5>

In [3]:
#1-(1)
A = np.array([ [1,3],[4,2] ])
A_values, A_vectors = np.linalg.eig(A)
print("numpy library")
print(" eigenvalues :",A_values, "\n","eigenvectors :", A_vectors)

numpy library
 eigenvalues : [-2.  5.] 
 eigenvectors : [[-0.70710678 -0.6       ]
 [ 0.70710678 -0.8       ]]


<h5>2번</h5>

<h3>제 개발환경에서 <code>init_printing()</code>이 작동하지 않습니다. <br>모듈 업데이트도 확인하고 VS코드나 python IDLE shell로도 해보았는데 작동이 되질 않습니다.. <br>혹시 작동 안하더라도.. 양해 부탁드립니다. </h3>

In [4]:
#1-(2)
A = Matrix([[1,3],[4,2]])
print("sympy library")
print(" eigenvalues :",A.eigenvals(),"\n","eigenvectors :", A.eigenvects())


sympy library
 eigenvalues : {5: 1, -2: 1} 
 eigenvectors : [(-2, 1, [Matrix([
[-1],
[ 1]])]), (5, 1, [Matrix([
[3/4],
[  1]])])]


<h5>3번</h5>

In [5]:
#1-(3)
t = symbols('t')
print(A.charpoly(t).as_expr())


t**2 - 3*t - 10


<h3>문제 2</h3>

In [6]:
#2
A = np.array([ [3,2,1],[2,1,4],[1,4,2] ])
values, vectors = np.linalg.eig(A)

P = vectors #고유벡터 P
P_inv = np.linalg.inv(P) #고유벡터 P의 역행렬
D = np.diag(values) #고유값을 대각행렬로 갖는 D

print("-----P-----\n", P) #가역행렬 P

#A = PDP^(-1)
result = np.matmul(P,D)
result = np.matmul(result,P_inv)

#같은지 확인
print("-----PDP^(-1)-----\n",result)
print("-----A-----\n",A)

#D = PAP^(-1)
result_2 = np.matmul(P_inv,A)
result_2 = np.matmul(result_2,P)
result_2 = result_2.astype(int)

#같은지 확인
print("-----PAP^(-1)-----\n",result_2)
print("-----D-----\n",D)


-----P-----
 [[ 0.49473045  0.85407884  0.16059614]
 [ 0.60797565 -0.20810216 -0.76619782]
 [ 0.62097294 -0.47669994  0.62221361]]
-----PDP^(-1)-----
 [[3. 2. 1.]
 [2. 1. 4.]
 [1. 4. 2.]]
-----A-----
 [[3 2 1]
 [2 1 4]
 [1 4 2]]
-----PAP^(-1)-----
 [[ 6  0  0]
 [ 0  1  0]
 [ 0  0 -2]]
-----D-----
 [[ 6.71297995  0.          0.        ]
 [ 0.          1.95454118  0.        ]
 [ 0.          0.         -2.66752113]]


<h5> $ D = P^{-1}AP $ , 이 되도록 하는 $D = \begin{pmatrix} 6.71297995 & 0 & 0 \\ 0 & 1.95454118 & 0 \\ 0 & 0 & -2.66752113 \end{pmatrix}$  $P = \begin{pmatrix} 0.49473045 & 0.85407884 & 0.16059614 \\ 0.60797565 & -0.20810216 & -0.76619782 \\ 0.62097294 & -0.47669994 & 0.62221361 \end{pmatrix}$ </h5>

<h3>문제 3</h3>

In [7]:
#3
A = np.array([ [1,1,1],[3,1,-3],[1,-2,-5] ])
P, L, U = lu(A)

# L, U 확인
print("----------L----------\n",L) #상삼각행렬
print("----------U----------\n",U) #하삼각행렬

#같은지 확인
print("-----LU-----\n",np.matmul(np.matmul(P,L),U))
print("-----A-----\n",A)

----------L----------
 [[ 1.          0.          0.        ]
 [ 0.33333333  1.          0.        ]
 [ 0.33333333 -0.28571429  1.        ]]
----------U----------
 [[ 3.          1.         -3.        ]
 [ 0.         -2.33333333 -4.        ]
 [ 0.          0.          0.85714286]]
-----LU-----
 [[ 1.  1.  1.]
 [ 3.  1. -3.]
 [ 1. -2. -5.]]
-----A-----
 [[ 1  1  1]
 [ 3  1 -3]
 [ 1 -2 -5]]


<h5> $L = \begin{pmatrix} 1 & 0 & 0 \\ 0.33333333 & 1 & 0 \\ 0.33333333 & -0.28571429 & 1 \end{pmatrix}$
     $U = \begin{pmatrix} 3 & 1 & -3 \\ 0 & -2.33333333 & -4 \\ 0 & 0 & 0.85714286 \end{pmatrix}$
     $LU = \begin{pmatrix} 1 & 1 & 1 \\ 3 & 1 & -3 \\ 1 & -2 & -5 \end{pmatrix}$
     $A = \begin{pmatrix} 1 & 1 & 1 \\ 3 & 1 & -3 \\ 1 & -2 & -5 \end{pmatrix}$
</h5>